sdv/metadata/single_table.py

```diff
@@ -307,15 +340,81 @@ def set_sequence_index(self, column_name):
                 Name of the sequence index column.
         """
-        if not isinstance(column_name, str):
-            raise ValueError("'sequence_index' must be a string.")
+        self._validate_sequence_index(column_name)
+        self._sequence_index = column_name
+
+    def _validate_sequence_index_not_in_sequence_key(self):
+        """Check that ``_sequence_index`` and ``_sequence_key`` don't overlap."""
+        sk = self._sequence_key
+        sequence_key = {sk} if isinstance(sk, str) else set(sk)
+        if self._sequence_index in sequence_key or sk is None:
+            index = {self._sequence_index}
+            raise ValueError(
+                f"'sequence_index' and 'sequence_key' have the same value {index}."
+                ' These columns must be different.'
+            )
 
-        self._metadata['sequence_index'] = column_name
+    def _validate_constraint(self, constraint_name, **kwargs):
+        """Validate a constraint against the single table metadata.
+
+        Args:
+            constraint_name (string):
+                Name of the constraint class.
+            **kwargs:
+                Any other arguments the constraint requires.
+        """
+        try:
+            constraint_class = Constraint._get_class_from_dict(constraint_name)
+        except KeyError:
+            raise InvalidMetadataError(f"Invalid constraint ('{constraint_name}').")
+
+        constraint_class._validate_metadata(self, **kwargs)
+
+    def _append_error(self, errors, method, *args, **kwargs):
+        """Inplace, append the produced error to the passed ``errors`` list."""
+        try:
+            method(*args, **kwargs)
+        except ValueError as e:
+            errors.append(e)
+
+    def validate(self):
+        """Validate the metadata.
+
+        Raises:
+            - ``InvalidMetadataError`` if the metadata is invalid.
+        """
+        # Validate constraints
+        errors = []
+        for tuple in self._constraints:
+            constraint_name, kwargs = tuple
+            try:
+                self._validate_constraint(constraint_name, **kwargs)
+            except MultipleConstraintsErrors as e:
+                reformated_errors = '\n'.join(map(str, e.errors))
+                errors.append(reformated_errors)
+
+        # Validate keys
+        self._append_error(errors, self._validate_key, self._primary_key, 'primary')
+        self._append_error(errors, self._validate_key, self._sequence_key, 'sequence')
+        self._append_error(errors, self._validate_alternate_keys, self._alternate_keys)
+        self._append_error(errors, self._validate_sequence_index, self._sequence_index)
+        self._append_error(errors, self._validate_sequence_index_not_in_sequence_key)
+
+        # Validate columns
+        for column, kwargs in self._columns.items():
+            self._append_error(errors, self._validate_column, column, **kwargs)
+
+        if errors:
+            raise InvalidMetadataError(
```

Comment:
Not necessary for this PR but later I think we might want to make a custom error that takes a list of errors and joins them in the message. We're probably going to be doing that a lot in the future